### PR TITLE
Added translation support to ontotext-yasgui.

### PR DESCRIPTION
**Pull request by boyan-tonchev**
## What
Added translation support to ontotext-yasgui.

## Why
ontotext-yasgui-web-component clients needs to support internationalization.

## How
 - Implemented a translation service.
 - Added external configuration which can extends supported languages. The client can change some translation of given language or add a new language at all.
 - Yasgui, Yasqe and Yasr don’t support transaction by default. The library configurations are extended to support translation function which will be used for translation. When the configurations are prepared the "translate" method of TranslationService is passed as value.

### CODE REVIEW

File: Yasgui/packages/yasgui/src/index.ts [hunk 1]
```diff
@@ -44,6 +44,7 @@ export interface Config<EndpointObject extends CatalogueItem = CatalogueItem> {
   requestConfig: YasguiRequestConfig;
   contextMenuContainer: HTMLElement | undefined;
   nonSslDomain?: string;
+  translate: (key: string, _parameters?: { key: string; value: string }[]) => string;
```
Comment (svilenvelikov) [inline Yasgui/packages/yasgui/src/index.ts:47]: Use `Record<string, string>[]` instead of `{ key: string; value: string }[]`

**Comment by svilenvelikov** [comment]
You should separate the changes in the yasgui source code in its own branch and then create a patch under `/yasgui-patches/22-12-2022-add-translate-service.patch`

File: ontotext-yasgui-web-component/src/services/translation.service.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+   *     "btn.mode-yasgui": "Editor and results",
+   *     "btn.mode-yasr": "Results only",
+   *   }
+   *   fr: {
+   *     "tooltip.switch.orientation.vertical": "Basculer vers verticale voir",
+   *     "btn.mode-yasqe": "Éditeur seulement",
+   *   }
+   * }
+   */
+  addTranslations(translations: { locale: { key: string, value: string } }) {
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/src/services/translation.service.ts:35]: Use Record instead of `{ key: string, value: string }`

File: ontotext-yasgui-web-component/src/services/translation.service.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  translate(key: string, parameters?: TranslationParameter[]): string {
+    let translation = this.getTranslation(key);
+    translation = this.applyParameters(translation, parameters);
+    return translation ? translation : key;
+  }
+
+  private getTranslation(key: string): string {
+    let selectedLang = this.currentLang;
+    if (!this.bundle || !this.bundle[this.currentLang]) {
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/src/services/translation.service.ts:55]: Shouldn't this check be done in the setLanguage method above?

File: ontotext-yasgui-web-component/src/services/translation.service.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      console.warn('Missing locale file for [' + this.currentLang + ']');
+      selectedLang = DEFAULT_LANG;
+    }
+
+    let translation = this.bundle[selectedLang][key];
+    if (!translation) {
+      // Fallback to English
+      translation = this.bundle[DEFAULT_LANG][key];
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/src/services/translation.service.ts:63]: It's not guaranteed that the key will be present in the default language bundle, so we should return the key at the end.

Comment (boyan-tonchev) [reply ontotext-yasgui-web-component/src/services/translation.service.ts:63]: Yes, you are right, but this check is done at the end of  "translate" function.

Comment (svilenvelikov) [reply ontotext-yasgui-web-component/src/services/translation.service.ts:63]: hmm I now saw that you actually return the key in the `translate` method if no translation is returned here. I think it'd be cleaner if you do it here. Also the parameters can be applied here too. This would simplify the logic and instant checking if translation is present

File: cypress/e2e/languages.spec.cy.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import {LanguagesSteps} from '../steps/languages-steps';
+import {ToolbarPageSteps} from '../steps/toolbar-page-steps';
+import {QueryStubs} from '../stubs/query-stubs';
+import {YasqeSteps} from '../steps/yasqe-steps';
+import {YasrSteps} from '../steps/yasr-steps';
+import {YasguiSteps} from '../steps/yasgui-steps';
```
Comment (svilenvelikov) [inline cypress/e2e/languages.spec.cy.ts:6]: Remove unused imports

File: cypress/steps/yasqe-steps.ts [hunk 6]
```diff
@@ -12,7 +12,7 @@ export class YasqeSteps {
     }
 
     static getExecuteQueryButton() {
-        return cy.get('.yasqe_queryButton');
+        return cy.get('.yasqe_queryButton:visible');
```
Comment (svilenvelikov) [inline cypress/steps/yasqe-steps.ts:15]: Why is this needed?

File: ontotext-yasgui-web-component/src/models/yasgui-configuration.ts [hunk 7]
```diff
@@ -40,6 +60,7 @@ export interface YasguiConfiguration {
    * The default yasgui config.
    */
   yasguiConfig: {
+    translate: (key: string, _parameters?: { key: string; value: string }[]) => string;
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/src/models/yasgui-configuration.ts:63]: Use Record or some custom type here too!

File: ontotext-yasgui-web-component/src/models/yasgui-configuration.ts [hunk 8]
```diff
@@ -30,6 +30,26 @@ export interface YasguiConfiguration {
...
+   *     "btn.mode-yasgui": "Editor and results",
+   *     "btn.mode-yasr": "Results only",
+   *   }
+   *   fr: {
+   *     "tooltip.switch.orientation.vertical": "Basculer vers verticale voir",
+   *     "btn.mode-yasqe": "Éditeur seulement",
+   *   }
+   * }
+   */
+  i18n?: {locale: Record<string, string>};
```
Comment (svilenvelikov) [inline ontotext-yasgui-web-component/src/models/yasgui-configuration.ts:51]: I see what `locale` means but I think its not well expressed and used. IMO better create some custom types like
```
export type Location = Record<string, string>;
export type Translations = Record<string, Location>;
```
and use them appropriately

**Comment by svilenvelikov** [comment]
I can't see why the package-lock is changed. Probably you should revert it

**Review by svilenvelikov — CHANGES_REQUESTED** [review]

File: cypress/steps/yasqe-steps.ts (hunk 6 above)
Comment (boyan-tonchev) [reply cypress/steps/yasqe-steps.ts:15]: I had a problem with button selection when second tab is opened. I thought that selector finds the hidden button from first tab. I removed it now and test passed successfully, I will remove it.

**Review by svilenvelikov — APPROVED** [review]
